Fixes #976: Resolve ambiguous constructors

mureinik opened the pull request:
With the current code (introduced in Mockito 2.7.14 by commit 6a82c03), calling `MockSettings.useConstructor` with an argument list that would be applicable to more than one constructor would fail with an `org.mockito.internal.creation.instance.InstantiationException`.

This behavior, however, is suboptimal, as described in issue #976, as it makes `useConstructor` less robust than the Java compiler, which is able to resolve such ambiguities.
With this patch, Mockito will attempt to match the constructor with the most specific parameter types. A constructor X is considered more specific than a constructor Y if:

1. They are both applicable to the given argument list
2. Constructor X has at least one parameter which is a further specialization of the corresponding parameter of constructor Y (i.e. `paramX.isAssignableFrom(paramY)`).
3. Constructor Y has no parameter which is a further specialization of the corresponding parameter of constructor X, as defined above.

E.g., consider the following class:

    public class SomeClass {
        SomeClass(Object o) {}
        SomeClass(String s) {}
    }

Without this patch, calling

    mock(SomeClass.class, withSettings().useConstructor("string!"))

would fail. With this patch, such a call would invoke the `SomeClass(String)` constructor.


As noted above, this PR fixes issue #976.

mureinik commented:
I wonder why Codecov has no previous report to compare against. Do any of the maintainers have a clue?

TimvdLippe commented:
@mureinik Because commits by the delivery drone do `[ci skip]` and therefore do not trigger a code coverage report.

mureinik commented:
@TimvdLippe That means that any PR that's rebased on top of a release won't have the benefit of a comparative CodeCov report. I get why the delivery drone should skip CI, but this is an unfortunate side effect. I'll bring this up on the mailing list - this PR probably isn't the right place to discuss it.

And thanks for the explanation!

TimvdLippe commented:
@mureinik please see #493 for reference.

mureinik commented:
@szczepiq ping? Will you have some time to take a look at this please?

szczepiq commented:
@mureinik, I'm very keen to review this change. Thank you for your submission!

I'm allocating most of my time to https://github.com/mockito/mockito/issues/911 currently but I promise to review (and hopefully merge) by the end of this weekend.

Thanks for patience with us :)

mureinik commented:
Rebased on top of 2.7.17 and added `CreatingMocksWithConstructorTest#can_spy_ambiguius_constructor_with_primitive` as per @TimvdLippe's suggestion above.

Note that this API does support implicit promotions (e.g., passing an `int` in `useConstructors()` varargs will autobox it to an `Integer`. This integer can't be matched to a constructor expecting a `long` primitive or a `Long` wrapper), in the same fashion that the `java.lang.reflect` APIs don't support it.

szczepiq commented on src/main/java/org/mockito/internal/creation/instance/ConstructorInstantiator.java:
```diff
@@ -130,4 +127,32 @@ private static boolean paramsMatch(Class<?>[] types, Object[] params) {
         return true;
     }
+
+    private void evaluateConstructor(List<Constructor<?>> matchingConstructors, Constructor<?> constructor) {
```
Can you add a javadoc describing this method? It has non trivial complexity and it would be handy to understand why we are implementing the algorithm this way.

mureinik replied:
Good point, thanks.
I'll add some javadoc to explain this.

szczepiq commented on src/test/java/org/mockitousage/constructor/CreatingMocksWithConstructorTest.java:
```diff
@@ -137,45 +137,90 @@ public void exception_message_when_constructor_not_found() {
     static class ExtendsExtendsBase extends ExtendsBase {}
```
I love the new tests. Thank you for adding comprehensive coverage!!!

szczepiq commented on src/main/java/org/mockito/internal/creation/instance/ConstructorInstantiator.java:
```diff
@@ -107,10 +107,7 @@ private String constructorArgsString() {
-                "If you believe that Mockito could do a better join deciding on which constructor to use, please let us know.",
-                "Ticket 685 contains the discussion and a workaround for ambiguous constructors using inner class.",
-                "See https://github.com/mockito/mockito/issues/685"
+                join("", " - ", constructors)
```
Let's keep the extra information. I think it is useful to:
 - inform about the inner class pattern for working around the ambiguous constructors
 - ask for feedback - it does not hurt to do that

You can keep the original or modify it if you want. I'd like to keep at least the hint about inner class.

mureinik replied:
I originally took this message as a "workaround" until this PR is implemented, but you're right, let's keep it. Note there's a typo there ("join" instead of "job"). I'll fix that while I'm at it.

mureinik commented:
Pushed updated branch. Change from previous push:

- Returned the error message regarding issue #685, as per @szczepiq review above
- Fixed a Freudian slip in the aforementioned error message ("a better **join**" -> "a better **job**")
- Added javadoc explaining the logic of `org.mockito.internal.creation.instance.ConstructorInstantiator#evaluateConstructor`, as per @szczepiq  review.

mureinik commented:
The build failure seems to be a TravisCI issue, not an issue with the code itself, as this revision introduced only minor changes (added javadoc and fixed a typo) compared to the previous one, and some of the Travis checks are passing. 

The error:

```
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
The build has been terminated
```

@szczepiq @TimvdLippe Is there any way to re-trigger Travis on this PR?

TimvdLippe commented:
I have restarted the builds
